### PR TITLE
Refactor operators using super and reshaped

### PR DESCRIPTION
**Pull request by cako**
## Description

This PR refactors linear operators based on the new pattern of calling `super().__init__` as soon as `dtype`, `dims`, `dimsd`, and other relevant parameters are defined, as introduced in #365. In addition, it also refactors many operators to rely on the `reshaped` decorator, introduced in #367.

Other changes: `Pad` and `DWT` always reshape internally (simplifies the code); `FirstDerivative` and `SecondDerivative` offload registering methods based on `kind` onto a separate method; `ChirpRadon2D`, `ChirpRadon3D` and `Fredholm1` now have `dims` and `dimsd`, allowing them to work with nd-array multiplication.

### CODE REVIEW

**Comment by cako** [comment]
@mrava87, I tried creating a new PR for the updates to the docs but it actually depends on this one. I will hold off submitting my next PR until this is resolved. But no need to hurry, take your time!

File: pylops/basicoperators/Diagonal.py [hunk 1]
```diff
@@ -63,42 +64,33 @@ class Diagonal(LinearOperator):
     """
 
     def __init__(self, diag, dims=None, axis=-1, dtype="float64", name="D"):
-        ncp = get_array_module(diag)
```
[the remark this answers is not in the document]
Comment (cako) [reply pylops/basicoperators/Diagonal.py:66]: Yea, some code got repeated somehow. Got rid of that.

Regarding the order of `super`, I have tried to ensure some consistency based on the principle of calling super as soon as it is possible, that is, as soon as all of its arguments are present. The idea is that someone reading the code will be able to separate what is the "base" code required for establishing `dtype`, `shape`, `dims`, `dimsd`, `clinear` and `explicit`, and what is specific to that operator. So these will be things like `axis`, `diag`, etc. You can see this design in other operators like `Flip`, `Identity`, `Imag`, `Real`, `Restriction`.

File: pylops/utils/decorators.py [hunk 2]
```diff
@@ -93,7 +95,11 @@ def _matvec(self, x):
     """
 
     def decorator(f):
-        forward = "rmat" not in f.__name__
+        forward = (
```
[the remark this answers is not in the document]
Comment (cako) [reply pylops/utils/decorators.py]: Fixed

File: pylops/basicoperators/Diagonal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrava87) [reply pylops/basicoperators/Diagonal.py:66]: Right, I agree :)

**Comment by mrava87** [comment]
If you think it’s ready, feel free to merge :)

**Comment by cako** [comment]
Could you approve :)

**Review by mrava87 — APPROVED** [review]

**Pull request merged**